compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp

```diff
@@ -222,25 +244,133 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
   return copyToWorkgroupMemory;
 }
 
+/// Return true if the op is used as operand `index` of an mma.sync op.
+static bool isMMAOperand(Operation* op, int64_t index) {
+  OpOperand* use = &(*((op->getUses()).begin()));
+  if (auto extract = dyn_cast<vector::ExtractStridedSliceOp>(use->getOwner())) {
+    use = &(*((extract->getUses()).begin()));
+  }
+  if (!isa<nvgpu::MmaSyncOp>(use->getOwner())) return false;
+  return use->getOperandNumber() == index;
+}
+
+/// Return true if the op is used as operand A of an mma.sync op.
+static bool isAOperand(Operation* op) { return isMMAOperand(op, 0); }
+/// Return true if the op is used as operand B of an mma.sync op.
+static bool isBOperand(Operation* op) { return isMMAOperand(op, 1); }
+
+/// Return a pipelining schedule that gives good performance on Nvidia
+/// Ampere target.
+static void getNvidiaTensorCorePipeline(
+    scf::ForOp forOp, std::vector<std::pair<Operation*, unsigned>>& ops,
+    unsigned depth) {
+  bool loopCanBePipelined = false;
+  // TODO: Tune this and make it a more fine grain logic.
+  static constexpr int64_t prefetchSize = 4;
+  SmallVector<Operation*> stage0;
+  SmallVector<Operation*> stagePrefetch;
+  SmallVector<Operation*> stageLast;
+  int64_t sizeReadA = 0;
```

Comment:
`sizeReadA` is bit miss-leading, `numLoadA`? Also, we should really be computing `prefetchSize` for different TileShape and InstructionShape. Also, `numPrefetchSmemLoadPerOperand`

Reply:
`prefetchSize` or `numPrefetchSmemLoadPerOperand`, if I get it right, is really the k-group worth of data being loaded from SMEM into Registers and we need compute is separately for `operandA` and `operandB`. For F16, this number is the same ldmatrix or ldmatrix.trans, but it will be different for F32 with ldmatrix and ld.shared. Also, as mentioned above the `numPrefetchSmemLoadPerOperand` will also change InstructionShape

Reply:
yes, we need a lot of fine tuning here. Improved the naming here and based on comment below.

Reply:
OK. So this will work for `mma.sync.m16n8k16.row.col` F16 data type for now and we can compute it later.

Reply:
Yes, I don't have enough data to handle the f32 case at this point. It should still be correct but perf may not be great